src/extension/api/configuration.ts

```diff
@@ -0,0 +1,64 @@
+import { BehaviorSubject } from 'rxjs'
+import { filter } from 'rxjs/operators'
+import * as sourcegraph from 'sourcegraph'
+import { ClientConfigurationAPI } from '../../client/api/configuration'
+import { ConfigurationCascade } from '../../protocol'
+
+/**
+ * @internal
+ * @template C - The configuration schema.
+ */
+class ExtConfigurationSection<C> implements sourcegraph.Configuration<C> {
+    constructor(private proxy: ClientConfigurationAPI, private data: C) {}
+
+    public get<K extends keyof C>(key: K): C[K] | undefined {
+        return this.data[key]
+    }
+
+    public async update<K extends keyof C>(key: K, value: C[K] | undefined): Promise<void> {
+        await this.proxy.$acceptConfigurationUpdate({ path: [key as string | number], value })
+    }
+
+    public get value(): Readonly<C> {
+        return this.data
+    }
+
+    public toJSON(): any {
+        return this.data
+    }
+}
+
+/**
+ * @internal
+ * @template C - The configuration schema.
+ */
+export interface ExtConfigurationAPI<C> {
+    $acceptConfigurationData(data: C): void
+}
+
+/**
+ * @internal
+ * @template C - The configuration schema.
+ */
+export class ExtConfiguration<C extends ConfigurationCascade<any>> implements ExtConfigurationAPI<C> {
+    private data = new BehaviorSubject<C | null>(null)
+
+    constructor(private proxy: ClientConfigurationAPI) {}
+
+    public $acceptConfigurationData(data: C): void {
+        this.data.next(Object.freeze(data))
```

Comment:
Why freeze this object at runtime but not make it readonly at compile time?

Reply:
I just added more `Readonly<>`s